internal/api/scim/resources/bulk.go

```diff
@@ -0,0 +1,264 @@
+	}
+	return request, nil
+}
+
+func (h *BulkHandler) processRequest(ctx context.Context, req *BulkRequest) (*BulkResponse, error) {
+	errorBudget := math.MaxInt32
+	if req.FailOnErrors != nil {
+		errorBudget = *req.FailOnErrors
+	}
+
+	resp := &BulkResponse{
+		Schemas:    []schemas.ScimSchemaType{schemas.IdBulkResponse},
+		Operations: make([]*BulkResponseOperation, 0, len(req.Operations)),
+	}
+
+	for _, operation := range req.Operations {
+		opResp := h.processOperation(ctx, operation)
+		resp.Operations = append(resp.Operations, opResp)
+
+		if opResp.Error == nil {
+			continue
+		}
+
+		errorBudget--
+		if errorBudget <= 0 {
+			return resp, nil
+		}
+	}
+
+	return resp, nil
+}
+
+func (h *BulkHandler) processOperation(ctx context.Context, op *BulkRequestOperation) (opResp *BulkResponseOperation) {
+	var statusCode int
+	var resourceNamePlural schemas.ScimResourceTypePlural
+	var resourceID string
+	var err error
+	opResp = &BulkResponseOperation{
+		Method: op.Method,
+		BulkID: op.BulkID,
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			logging.WithFields("panic", r).Error("Bulk operation panic")
+			err = zerrors.ThrowInternal(nil, "SCIM-BLK12", "Internal error while processing bulk operation")
+		}
+
+		if resourceNamePlural != "" && resourceID != "" {
+			opResp.Location = buildLocation(ctx, resourceNamePlural, resourceID)
+		}
+
+		opResp.Status = strconv.Itoa(statusCode)
+		if err != nil {
+			opResp.Error = serrors.MapToScimError(ctx, err)
+			opResp.Status = opResp.Error.Status
+		}
+	}()
+
+	resourceNamePlural, resourceID, err = h.parsePath(op.Path)
+	if err != nil {
+		return opResp
+	}
+
+	resourceID, err = metadata.ResolveScimBulkIDIfNeeded(ctx, resourceID)
+	if err != nil {
+		return opResp
+	}
+
+	resourceHandler, ok := h.handlersByPluralResourceName[resourceNamePlural]
+	if !ok {
+		err = zerrors.ThrowInvalidArgumentf(nil, "SCIM-BLK13", "Unknown resource %s", resourceNamePlural)
+		return opResp
+	}
+
+	switch op.Method {
+	case http.MethodPatch:
+		statusCode = http.StatusNoContent
+		err = h.processPatchOperation(ctx, resourceHandler, resourceID, op.Data)
+	case http.MethodPut:
+		statusCode = http.StatusOK
+		err = h.processPutOperation(ctx, resourceHandler, resourceID, op)
+	case http.MethodPost:
+		statusCode = http.StatusCreated
+		resourceID, err = h.processPostOperation(ctx, resourceHandler, resourceID, op)
+	case http.MethodDelete:
+		statusCode = http.StatusNoContent
+		err = h.processDeleteOperation(ctx, resourceHandler, resourceID)
+	default:
+		err = zerrors.ThrowInvalidArgumentf(nil, "SCIM-BLK14", "Unsupported operation %s", op.Method)
+	}
+
+	return opResp
+}
+
+func (h *BulkHandler) processPutOperation(ctx context.Context, resourceHandler RawResourceHandlerAdapter, resourceID string, op *BulkRequestOperation) error {
+	if err := h.checkPermission(ctx, domain.PermissionUserWrite, resourceID); err != nil {
```

Comment:
I'm uncertain about the details of these permission checks and the underlying concept behind them. Could you clarify when permissions are evaluated at the command/query layer versus when they are enforced at the interface layer?

Reply:
Command layer has the permission checks included, at least related to the creation and update of a user. Same for the query layer, related to what the executing user is able to see.

Reply:
Which should mean that you would not need any permission check in this location.

Reply:
I revisited this and believe I now understand it better. I’ve removed the checks in the bulk layer. Initially, I added these checks to ensure that the calling user has permission not only on the target resource but also on the organization specified in the SCIM URL.
However, I think this raises a broader discussion. Currently, there is no enforcement that a target user's resource owner must match the organization in the SCIM URL path. For example, if a machine user has permissions to delete users in both OrgA and OrgB, and user1’s resource owner is OrgA, but the machine user sends a delete request to OrgB’s SCIM endpoint, the deletion still succeeds. Shouldn't this instead result in a "not found" error? What do you think? If we decide to implement this, I'd do so in another PR.

Reply:
The permission check includes the check on the specific resource and the resourceowner on the command side.
You can add the resourceowner to the UserHumanWriteModel as a parameter, if then the user is under another resourceowner and you want to update it, you get back a NotFound-error.
Should be possible in this PR, or do you think it takes a lot of effort?

Reply:
https://github.com/zitadel/zitadel/blob/accfb7525a63ad9b3b14d6ad315b3d05a59f0ff3/internal/command/user_v2_human.go#L510

Reply:
I would handle this separately since it likely affects unrelated endpoints (update, replace, get, etc.) and requires tests for all of them (quite unrelated to the bulk endpoint).

Reply:
Sounds good to me, then let's do a separate PR as a "fix", to add this and the related tests.